stack.hpp

```diff
@@ -16,17 +21,17 @@ class Stack {
     }
 
     void push(const T& element) {
-        if (position_ > capacity_) throw std::out_of_range("Not enough capacity");
-        storage_[position_++] = element;
+        if (position_ >= capacity_) throw std::out_of_range("Not enough capacity");
+        storage_[++position_] = element;
     }
 
     void push(T&& element) {
-        if (position_ > capacity_) throw std::out_of_range("Not enough capacity");
-        storage_[position_++] = std::move(element);
+        if (position_ >= capacity_) throw std::out_of_range("Not enough capacity");
+        storage_[++position_] = std::move(element);
     }
 
     T pop() {
-        if (position_ < 0) throw std::out_of_range("Stack is empty");
+        if (position_ <= 0) throw std::out_of_range("Stack is empty");
         return std::move(storage_[position_--]);
```

Comment:
The problem is rather here: position_ points to the location where a new element can be inserted, so the last element is one position before...